### ISSUE TITLE
PlatyBrowser [level] button is missing

### ISSUE THREAD

**Issue by tischi** [mobie/mobie-viewer-fiji#532]

**Comment by tischi** [comment]
@K-Meech @constantinpape 

Can you help me find a commit where we still had a leveling button?

I was expecting that this should have it: https://github.com/mobie/mobie-viewer-fiji/releases/tag/v0.6.0

But searching for `createLevelingPanel` in this release on github showed that the method was not used. Maybe I am doing something wrong?

**Comment by tischi** [comment]
I looks like on github it is always searching the latest commit...

When I `git checkout tags/v0.6.0 -b v0.6.0` I could find the old code.

Should we hard code the leveling vector in Java or do we (should we) rather have some JSON field for this? I vaguely remember that we discussed this, but this is looooooong ago.

```
private double[] defaultTargetNormalVector = new double[]{0.70,0.56,0.43};
```

**Comment by tischi** [comment]
@K-Meech I am thinking of adding a new implementation of `public interface ViewerTransform`, namely a `NormalVectorViewerTransform` (better name?) with a field called `normalVector`. This could then be used with the `MoBIEViewerTransformChanger`.

In general we do have already a `ViewerTransformAdapter` that could then be used to `deserialize` such a `NormalVectorViewerTransform`.

--------------

Then maybe we do **not** add back the level button, but just add another view category here:

<img width="536" alt="image" src="https://user-images.githubusercontent.com/2157566/141825264-63c8e262-f4e6-41f0-b845-94cb4e282b39.png">


maybe: `levelling:   dorsal-ventral` and then we could add other normal vectors for other directions.
  
what do you guys think?

@HernandoMV

**Comment by constantinpape** [comment]
The leveling vector is already saved in `misc`: https://github.com/mobie/platybrowser-project/blob/main/data/1.0.1/misc/leveling.json.

**Comment by constantinpape** [comment]
> I am thinking of adding a new implementation of `public interface ViewerTransform`, namely a `NormalVectorViewerTransform` (better name?) with a field called `normalVector`. This could then be used with the `MoBIEViewerTransformChanger`.
> 
> In general we do have already a `ViewerTransformAdapter` that could then be used to `deserialize` such a `NormalVectorViewerTransform`.

I agree that this makes more sense than having the special functionality for leveling. Should I add it to the spec and create the corresponding view for the platybrowser?

**Comment by tischi** [comment]
> I agree that this makes more sense than having the special functionality for leveling. Should I add it to the spec and create the corresponding view for the platybrowser?

Let's maybe wait for @K-Meech feedback. 

Is `NormalVectorViewerTransform` the best name? I guess the point is that the transform would change the viewer such that the viewer z-axis is parallel to the given vector.

**Comment by constantinpape** [comment]
> Is `NormalVectorViewerTransform` the best name?

I think as a Java name that sounds good. In the spec I would call it `normalVector`.

**Comment by K-Meech** [comment]
Yes - this sounds good! Makes a lot more sense than having specific leveling stuff for the platybrowser only. I guess you would need to support having an optional position + timepoint too, otherwise there's no way to specify where the viewer ends up exactly.

**Comment by tischi** [comment]
OK, I will work on it now. Maybe we call the new views: `anatomy`? How should we call the "level" view? Something with bilateral symmetry or rather some dorsal ventral apical basal stuff? We could then also add views like: gut, head, ....

**Comment by K-Meech** [comment]
Sounds good! Probably something like transverse/ sagittal/coronal plane... Although I can never remember which is which..

**Comment by tischi** [comment]
@K-Meech I am not sure we need the position, e.g. we also have:

```
public class TimepointViewerTransform implements ViewerTransform
{
	// Serialization
	private Integer timepoint;

	public TimepointViewerTransform( int timepoint )
	{
		this.timepoint = timepoint;
	}

	@Override
	public double[] getParameters()
	{
		return null;
	}

	@Override
	public Integer getTimepoint()
	{
		return timepoint;
	}
}
```

**Comment by tischi** [comment]
@constantinpape Could please add a view that is using this?

```
// JSON Adapter
nameToClass.put("normalVector", NormalVectorViewerTransform.class);
classToName.put(NormalVectorViewerTransform.class.getName(), "normalVector");
```

```
public class NormalVectorViewerTransform implements ViewerTransform
{
	// Serialisation
	private double[] normalVector; // required
	private final Integer timepoint; // optional
```

**Comment by HernandoMV** [comment]
I agree with Kimberly. I think "Anatomical views" is the best term for it, and we can have (at least) 4 possibilities:
- The 'level' that we had, that divides the body into front and back: CORONAL
- An anterio-posterior plane, that divides the body into upper and lower: AXIAL 
- A left-right division of the body: SAGITAL
- Imaging plane

**Comment by K-Meech** [comment]
> @K-Meech I am not sure we need the position, e.g. we also have:

@tischi Yes it's not totally necessary for sure! It just might be convenient if say you had a normal vector for a certain orientation (like a sagital section), and you wanted to see that through a specific cell at a specific position. Currently, you'd have to go there and save that as the usual normalisedAffine view.

**Comment by tischi** [comment]
But then it still would not zoom onto this cell, and if we also specify the zoom then we are back at a full affine...

**Comment by constantinpape** [comment]
I wouldn't add position or timepoint for this one; I think the idea is just to set the viewer plane so that it's orthogonal to the normal vector.

**Comment by K-Meech** [comment]
Ok - no worries! We can just leave it as the normal vector for now. I think it would only be for particular use cases - where you want to see a specific **plane** (i.e. normal and point), rather than only a specific  **normal** - where it would be useful.

**Comment by constantinpape** [comment]
> I think it would only be for particular use cases - where you want to see a specific **plane**

But I think for that one we can just use normalizedAffine or affine (sure that also includes the zoom, but I think that's probably ok in most use cases where you want to go somewhere specific)

**Comment by K-Meech** [comment]
Yes - for sure! Sounds good. Using the affine makes sense. (I'm biased because I often calculate the normal/point of planes I want to get to in bigdataviewer, and converting that to a corresponding affine is a bit of a pain. It's a super niche use case though - I agree!)

**Comment by tischi** [comment]
On the Java side we are currently stuck with `timepoint` because: 
```
public interface ViewerTransform
{
	double[] getParameters();
	Integer getTimepoint();
}
```
But I checked the code: if timepoint is NULL it will simply stay at the current timepoint.

**Comment by constantinpape** [comment]
> On the Java side we are currently stuck with `timepoint` because:

Ok, I will add it as an optional field then.

**Comment by constantinpape** [comment]
@tischi https://github.com/mobie/platybrowser-project/pull/88
It has "anatomical-views"->"coronal".

**Comment by tischi** [comment]
what does "coronal" mean?

I think "anatomy" is nicer/conciser than "anatomical-views" (those are all views, so this feels redundant).

**Comment by constantinpape** [comment]
> what does "coronal" mean?
> 
> I think "anatomy" is nicer/conciser than "anatomical-views" (those are all views, so this feels redundant).

See the comment by @HernandoMV: https://github.com/mobie/mobie-viewer-fiji/issues/532#issuecomment-970057437

But feel free to change the names as you like, you have write access to the repo.

**Comment by tischi** [comment]
> @tischi mobie/platybrowser-project#88
It has "anatomical-views"->"coronal".

Can we merge this? The code already should be able to handle it.

**Comment by constantinpape** [comment]
Ready to merge from my side as soon as it also works with the MoBIE from the main update site.
